### PR TITLE
(WIP) (PUP-6476) add hocon support to puppet

### PR DESCRIPTION
**Pull request by janelu2**
This PR adds backend support of HOCON files for Puppet Lookup. 
Changes made: 
- add HOCON gem requirement
- add a HoconDataProviderFactory and include it to the registry

### CODE REVIEW

**Comment by janelu2** [comment]
@cprice404 @thallgren currently fighting some strange windows appv failure, but that is related specifically to the utf8 encoding so the rest is fine. Give it a look?

File: Gemfile [hunk 1]
```diff
@@ -27,6 +27,7 @@ gem "puppet", :path => File.dirname(__FILE__), :require => false
 gem "facter", *location_for(ENV['FACTER_LOCATION'] || ['> 2.0', '< 4'])
 gem "hiera", *location_for(ENV['HIERA_LOCATION'] || ['>= 2.0', '< 4'])
 gem "rake", "10.1.1", :require => false
+gem 'hocon', '~> 1.1.1'
```
Comment (cprice404) [inline Gemfile]: @kylog this is only used for development, right?  So it's OK to have this here and we will just, additionally, need to sort out the packaging issues we've been discussing?

File: spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_hocon/data/bad.hocon [hunk 2]
```diff
@@ -0,0 +1 @@
+"test::param_a": "env data param_a is 10":
```
Comment (cprice404) [inline spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_hocon/data/bad.hocon]: Is there any meaning to the fact that this file is named `bad`, or is the file name just arbitrary?

**Comment by cprice404** [comment]
+1 other than the windows failure and the need to resolve the gem packaging issue.

File: spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_hocon/data/bad.hocon (hunk 2 above)
Comment (janelu2) [inline spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_hocon/data/bad.hocon]: this is a hocon file with a syntax error (the colon at the end) hence the "bad"

Comment (cprice404) [inline spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_hocon/data/bad.hocon]: d'oh, i didn't notice the colon.  that makes sense.

File: Gemfile (hunk 1 above)
Comment (janelu2) [inline Gemfile]: temporary updated this line to point to @Iristyle's PR, will change it back once a new release comes out with his fix

**Comment by janelu2** [comment]
currently changed the hocon gem to be pointed to @Iristyle's PR for ruby-hocon, I will change that bundler hack after a new version of hocon gets released with the utf8 encoding patch

File: lib/puppet/data_providers/hocon_data_provider_factory.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    def path_extension
+      '.hocon'
+    end
+  end
+
+  class HoconDataProvider < Puppet::Plugins::DataProviders::PathBasedDataProvider
+    def initialize_data(path, lookup_invocation)
+      data = Hocon.load(File.absolute_path(path), {:syntax => Hocon::ConfigSyntax::HOCON})
+      HieraConfig.symkeys_to_string(data.nil? ? {} : data)
```
Comment (thallgren) [inline lib/puppet/data_providers/hocon_data_provider_factory.rb]: Will the Hocon parser produce keys that are ruby Symbols? The Hiera ruby implementation has a special interpretation for keys starting with ':' but that's not standard. This line should be removed unless the Hocon implementation has the same behavior,

Comment (janelu2) [inline lib/puppet/data_providers/hocon_data_provider_factory.rb]: just to be sure, you are referring to line 21 right?

Comment (janelu2) [inline lib/puppet/data_providers/hocon_data_provider_factory.rb]: FYI I've removed that line

**Comment by janelu2** [comment]
woop woop hocon version updated to v1.1.2 to support utf8 issues with windows

**Comment by hlindberg** [comment]
Ah, this is superseded by PR  #5463

**Comment by hlindberg** [comment]
Closing this in favor of #5463

**Pull request closed without merging**